### PR TITLE
Adding filter option to phel test command

### PR DESCRIPTION
**Pull request by Chemaclass**
## 📚 Description

Adding a new concept to the phel commands: "options". With this idea, we can extend the functionality of the different commands that we already have in a simple way.

The command arguments get divided by "inputs" and "options". The difference is basically that the options start with `--`

## 🔖 Changes

- Added an option `--filter=...` to the Test command

### CODE REVIEW

File: src/phel/test.phel [hunk 1]
```diff
@@ -298,19 +302,21 @@
           :when (get meta :test)]
       (get-in php/$GLOBALS ["__phel" munge-ns fn-name]))))
 
-(defn- run-test [ns]
+(defn- run-test [options ns]
```
Comment (jenshaase) [inline src/phel/test.phel:305]: You can use the destructuring here `[{:filter filter-options} ns]`

File: src/phel/test.phel [hunk 2]
```diff
@@ -298,19 +302,21 @@
...
   (let [rt (php/:: \Phel\Runtime\RuntimeSingleton (getInstance))
         _ (php/-> rt (loadNs ns))
-        tests (find-test-fns ns)]
-    (dofor [test :in tests]
+        tests (find-test-fns ns)
```
Comment (jenshaase) [inline src/phel/test.phel:308]: I would do the following here:

1. Change `deftest` to `...(defn ,name {:test true :test-name ,name} ...` (I added `:test-name`)
2. Pass the `filter-options` to `find-test-fns` and do the filtering there. You get the test name with `(get meta :test-name)`

The `BOUND` variable is currently used for internals only (It is used to print better exceptions).

File: src/php/Command/Test/TestCommandOptions.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    public static function empty(): self
+    {
+        return new self();
+    }
+
+    public static function fromArray(array $options): self
+    {
+        $self = new self();
+
+        foreach ($options as $option) {
```
Comment (jenshaase) [inline src/php/Command/Test/TestCommandOptions.php:22]: I would recommend to use `getopt` (https://www.php.net/manual/de/function.getopt.php) or symfony/console  for parsing command line arguments. Everything else will probably fail (edge cases)

File: src/php/Command/Test/TestCommandOptions.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            if (false !== strpos($option, self::FILTER_OPTION)) {
+                $self->filter = str_replace('-', '_', substr($option, strlen(self::FILTER_OPTION)));
+            }
+        }
+
+        return $self;
+    }
+
+    public function asPhelHashMap(): string
```
Comment (jenshaase) [inline src/php/Command/Test/TestCommandOptions.php:31]: This looks ugly and there are probably some edge case where it will fail. Why not make a new instance of the Map class (using TypeFactory) and then use the Printer to print the result?

**Review by jenshaase — CHANGES_REQUESTED** [review]

File: src/php/Command/Test/TestCommandOptions.php (hunk 4 above)
Comment (Chemaclass) [reply src/php/Command/Test/TestCommandOptions.php:31]: I agree. Let's talk in private to agree on this. Let's sync first on the way of how do we want to handle inputs and options.

**Comment by Chemaclass** [comment]
After some brainstorming, we (@jenshaase and I) agreed on closing this PR in favor of:
1st PR: using `symfony/console` command for all commands
2nd PR: expand the commands themselves (adding options or whatever)
